cmd/goal/box.go

```diff
@@ -0,0 +1,117 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+var boxName string
+
+func init() {
+	appCmd.AddCommand(appBoxCmd)
+
+	appBoxCmd.AddCommand(appBoxInfoCmd)
+	appBoxCmd.AddCommand(appBoxListCmd)
+	appBoxCmd.PersistentFlags().Uint64Var(&appIdx, "app-id", 0, "Application ID")
+	appBoxCmd.MarkFlagRequired("app-id")
+
+	appBoxInfoCmd.Flags().StringVarP(&boxName, "name", "n", "", "Application box name. Use the same form as app-arg to name the box.")
+	appBoxInfoCmd.MarkFlagRequired("name")
+}
+
+var appBoxCmd = &cobra.Command{
+	Use:   "box",
+	Short: "Read application box data",
+	Args:  cobra.ArbitraryArgs,
+	Run: func(cmd *cobra.Command, args []string) {
+		// If no arguments passed, we should fallback to help
+		cmd.HelpFunc()(cmd, args)
+	},
+}
+
+var appBoxInfoCmd = &cobra.Command{
+	Use:   "info",
+	Short: "Retrieve information about an application box.",
+	Long:  "Retrieve information about an application box. The returned box name and value are base64 encoded.",
+	Args:  validateNoPosArgsFn,
+	Run: func(cmd *cobra.Command, args []string) {
+		_, client := getDataDirAndClient()
+
+		// Ensure box name is specified
+		if boxName == "" {
+			reportErrorf(errorMissingBoxName)
+		}
+
+		// Get box info
+		box, err := client.GetApplicationBoxByName(appIdx, boxName)
+		if err != nil {
+			if strings.Contains(err.Error(), fmt.Sprintf("{\"message\":\"box not found\"}")) {
+				reportErrorf("No box found for appid %d with name %s", appIdx, boxName)
+			}
+			reportErrorf(errorRequestFail, err)
+		}
+
+		// Print inputted box name, but check that it matches found box name first
+		// This reduces confusion of potentially receiving a different box name representation
+		boxNameBytes, err := newAppCallBytes(boxName).Raw()
+		if err != nil {
+			reportErrorf(errorInvalidBoxName, boxName)
+		}
+		if bytes.Compare(box.Name, boxNameBytes) != 0 {
+			reportErrorf(errorBoxNameMismatch, box.Name, boxNameBytes)
+		}
+		reportInfof("Name:  %s", boxName)
+
+		// Print box value
+		reportInfof("Value: %s", encodeBytesAsAppCallBytes(box.Value))
+	},
+}
+
+var appBoxListCmd = &cobra.Command{
+	Use:   "list",
+	Short: "List all application boxes belonging to an application",
+	Long: "List all application boxes belonging to an application.\n" +
+		"For printable strings, the box name is formatted as 'str:hello'\n" +
+		"For everything else, the box name is formatted as 'b64:A=='. ",
+	Args: validateNoPosArgsFn,
+	Run: func(cmd *cobra.Command, args []string) {
+		_, client := getDataDirAndClient()
+
+		// Get app boxes
+		boxesRes, err := client.ApplicationBoxes(appIdx)
```

Comment:
@ahangsu I believe that you created the receiving method `func (c *Client) ApplicationBoxes(index uint64) (resp generatedV2.BoxesResponse, err error)` in `libgoal/libgoal.go`. It doesn't expose the `max` parameter. I'm not sure that it should but I'm thinking about the case when someone has tons of boxes in an app, and wants to see what all of them are (so set `?max=0`). Perhaps in that case they can just configure their Node via `v2.Node.Config().MaxAPIBoxPerApplication = 0` ? I'm just wondering 

This also relates to the paging question _in indexer_ that @michaeldiamant brought to my attention.

Reply:
looking into it, I am opening a small PR to allow for passing in optional `max` box threshold, by passing in `*uint64` pointer, so hopefully this should help to give a versatile set for queries.

For node option, which is a hard limit on box-names to return, I think you might be right about the configuration, but that is kinda out of scope of the current problem I guess.

Reply:
This PR: https://github.com/algorand/go-algorand/pull/4268, may help passing in a limit of box names to return in query, if you are interested please consider take a quick glimpse.